Collision branch

CheezitBandit opened the pull request:
First commit for moving platforms - initialized in PlayState by ObjVar = new Platforms(startingXposition, startingYposition, widthOfBlock*, minX, maxX, minY, maxY); .

Width of block is requested as an integer, how many "blocks" of the standard 32x32 size are joined length-wise to create the entire moving platform.

MinX/Y and MaxX/Y are the starting positions +/- the min/max given.

CheezitBandit committed:
Moving platform, formatted to allow input for any platform to start (x, y) and move (L, R, U, D)

CheezitBandit committed:
Adding adjustable width functionality

CheezitBandit committed:
Added comments

CheezitBandit committed:
Corrected reference error that had been previously commented out (whoops)

CheezitBandit committed:
Player sprite now sticks to platform moving U/L/R, and sticky position is redefined on re-touch

SamBumgardner commented:
I'll look over this code and give it a proper review today before class begins.

CheezitBandit committed:
Sprite sticks U/L/R, redefines touch, does not interact with platform unless lands directly on top

CheezitBandit committed:
Sprite on platform responds to key presses. Still issues on platform dropping from beneath it.

CheezitBandit committed:
Sprite moves with platform, can jump down through platform on key press.

SamBumgardner commented:
Good work on getting the player to stick to the platform!
The code right now kinda messes with the player's x acceleration and velocity, though, because it only updates the platform offset if a movement key is held. This can cause weird situations when the player runs in one direction, lets go of all movement keys, then taps the opposite movement key.

I found that using the player's velocity (multiplied by elapsed time) to adjust the offset each frame instead of the difference between the player's current position and previous position works pretty well. I'll submit a pull request on your branch with the changes I made for you to look over.

SamBumgardner committed:
Changed the platform's movement to just use velocity.

This smooths out the movement, and makes it easier to adjust the speed of the platform.
It looked a bit jagged to me before.

SamBumgardner committed:
Changed platform offset updating logic.

This uses the player's velocity to adjust their platform offset every frame,
which means that the player properly accellerates and decelerates when
on the moving platform.

SamBumgardner commented:
Okay, I made a PR to your branch. Let me know there if you have any questions/comments/concerns!

I'll do a more detailed review later.

CheezitBandit committed:
Merge pull request #1 from SamBumgardner/rework-offset-incrementing

Reworked Offset Updating

CheezitBandit committed:
Merge branch 'master' of https://github.com/CheezitBandit/csc-303-platformer

# Conflicts:
#	source/PlayState.hx

CheezitBandit committed:
Merge branch 'Collision_Branch'

* Collision_Branch:
  Changed platform offset updating logic.
  Changed the platform's movement to just use velocity.
  Sprite moves with platform, can jump down through platform on key press.
  Sprite on platform responds to key presses. Still issues on platform dropping from beneath it.
  Sprite sticks U/L/R, redefines touch, does not interact with platform unless lands directly on top
  Player sprite now sticks to platform moving U/L/R, and sticky position is redefined on re-touch
  Corrected reference error that had been previously commented out (whoops)
  Added comments
  Adding adjustable width functionality

# Conflicts:
#	source/PlayState.hx - JLG manually merged files together, tested working result

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,121 @@
+    public var offsetX:Float = 0;
+    public var sticky:Bool = false;
+    public var newbool:Bool;
+	public function new(?X:Float=0, ?Y:Float=0, ?W:Int=0, ?L:Float=0, ?R:Float=0, ?U:Float=0, ?D:Float=0, ?trackPlayer:Player, ?SimpleGraphic:FlxGraphicAsset) 
```
This function should be un-indented by one level.

CheezitBandit replied:
Oddly enough, it's always looked fine from my view. I'll keep an eye on how it appears on github on the next commits.

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,121 @@
+            }
+            else {
+                velocity.y = 40;
+            }
```
The indentation here needs to be fixed up as well.

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,121 @@
+        maxX = R;
+        minY = U;
+        maxY = D;
+        platformwidth = 32 * W;
```
It'd be nice to create a read-only variable with the tile size in it instead of putting 32 here raw.

CheezitBandit replied:
Makes a lot of sense, very similar to the pulling gravity out and making it a variable instead of the plain number. The block dimension has been pulled out and made a separate variable, private var blockSize, which has been set at 32, same as the blocks on the map.

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,121 @@
+                DOWN = true;
+            }
+            else {
+                velocity.y = -40;
```
All of these `40`s could be replaced with a single variable.

CheezitBandit replied:
The new private variable platformVelocity has been created and set to a value of 40.

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,121 @@
+        }
+    }
+
+    //returns offset of player when player jumps on block
```
We shouldn't leave large swaths of commented-out code in any files we plan to merge into master. If this code is/can be useful, it should be uncommented and improved (if necessary). Otherwise, we should just delete everything.

CheezitBandit replied:
You're right, and I should know better. The commented out code has been removed.

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,121 @@
+	{
+        movement();
+        // movePlayer(player);
+        //trace("actually moving player");
```
Same thing as above. Remove or use, but don't leave these lines commented out.

SamBumgardner commented on source/PlayState.hx:
```diff
@@ -5,13 +5,17 @@ import flixel.FlxState;
-	private var player:Player;
+	public var player:Player;
+	private var platform:Platforms;
+	private var x:Float = 0;
```
I don't think this variable is being used for anything. If it isn't, then remove it.

SamBumgardner commented on source/PlayState.hx:
```diff
@@ -22,8 +26,17 @@ class PlayState extends FlxState
 		super.create();
 		
 		player = new Player(50, 50);
+		//player.collisonXDrag = true;
```
Can remove this line.

SamBumgardner commented on source/PlayState.hx:
```diff
@@ -22,8 +26,17 @@ class PlayState extends FlxState
+		//create new moving platform
+		platform =  new Platforms(250, 150, 3, 100, 100, 50, 50, player);
+		platform.immovable = platform.solid = true;
+		//platform.collisonXDrag = true;
```
Can remove this line.

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,121 @@
+    // private var player:Player;
+    private var tracer:Int = 0;
+    public var offsetX:Float = 0;
+    public var sticky:Bool = false;
```
Maybe consider renaming this variable. Sticky could work, but there may be another name that gets across its purpose (identifying if the player is in contact with the platform).

CheezitBandit replied:
That makes a lot of sense. The variable was renamed inContact.

SamBumgardner commented on source/PlayState.hx:
```diff
@@ -49,7 +62,37 @@ class PlayState extends FlxState
+		}
+		else {
+			player.acceleration.y = 600;
+		}
```
This whole section of player-platform collision logic should be moved out of playstate, if possible. 
Calls to collide and overlap will still be here, but the rest of this can be moved to functions inside your Platforms class.

CheezitBandit replied:
I've created a new function in Platforms.hx that checks for collisions and overlaps. It is called in the update function in PlayState.

SamBumgardner commented on source/PlayState.hx:
```diff
@@ -49,7 +62,37 @@ class PlayState extends FlxState
+			
+			// Allow player to drop through platform if down key pressed
+			if (FlxG.keys.anyPressed([FlxKey.DOWN])) {
+				player.y = platform.y;
```
I don't feel like this is the best way to make the player drop through a platform, because it jumps them down to being beneath the platform in a single frame.

Instead, I think you could only check for a collision between player and platform if down is **not** held, which means the player will just naturally drop through the platform if the player holds down.

CheezitBandit replied:
This has been addressed in the next commit. It was much easier to implement than I expected, and worked perfectly. Thank you for the suggestion.

SamBumgardner commented on source/PlayState.hx:
```diff
@@ -49,7 +62,37 @@ class PlayState extends FlxState
+
+		if (FlxG.collide(player, platform)) {
+			// Only set offset value if touching platform for "first" time
+			player.acceleration.y = 8000;
```
I'd rather have this number stored in a variable in the Platforms class, if possible.

CheezitBandit replied:
Done, it is now a variable called heavy.

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,121 @@
+    private var startY:Float;
+    private var platformwidth:Int;
+    // private var player:Player;
+    private var tracer:Int = 0;
```
`tracer` isn't being used at the moment, so probably remove it.

CheezitBandit replied:
Done

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,121 @@
+    private var tracer:Int = 0;
+    public var offsetX:Float = 0;
+    public var sticky:Bool = false;
+    public var newbool:Bool;
```
`newbool` isn't being used right now either, so probably remove it.

CheezitBandit replied:
Done

SamBumgardner reviewed (changes requested):
Overall, the code works pretty well! There are some issues that I made notes of with code comments, but it's mostly just cleanup stuff. 

I do want all of your functions to have the kind of documentation that the rest of the project has (see the player class for examples), but that's something that can be added on at the end if necessary.

The challenge of supporting multiple objects being on the platform is definitely more complex. Here are my thoughts on how to solve the problem of allowing multiple things to stand on top of the platform:

Right now, to make the player move along with the platform, you need to keep track of two variables for them:   `sticky` and `offsetX`. So, to keep track of multiple objects, you need to be able to maintain a sticky and offsetX variable for each of these objects, and do your offset and moving operations on each of them.

To do this, you could create a new class that bundles together a reference to an FlxObject (the thing on the platform) a `collidedThisTurn` variable, and `offsetX`. The moving platform will keep track of an array<new class>. This eliminates the need for `sticky`, I think, since being in the tracking array is like having sticky be true.

When an object collides with the moving platform, iterate through that tracking array, identify if the object is already in the tracking list. If it is in the tracking list, set `collidedThisTurn` to `true`. If it isn't in the tracking list, add an new entry to the tracking list, set up its offset, etc. and set `collidedThisTurn` to true. 

Then do all of the code you did for the payer to update the position of the object that is in contact with the moving platform.

At the end of an update cycle, (after all collisions have been resolved) iterate through the tracking array. Every `new class` object in the tracking array with `collidedThisTurn == true` is kept in the array, and `collidedThisTurn` is set to false. Every object where `collidedThisTurn == false` is removed from the tracking array using the `swapAndPop()` method found [here](http://api.haxeflixel.com/flixel/util/FlxArrayUtil.html#swapAndPop)

CheezitBandit committed:
First set of (mostly editorial) changes to files as requested in code review.

CheezitBandit committed:
Corrected sloppy implementation of jumping down through platform as suggested by Sam Bumgardner

CheezitBandit committed:
Created FlxTypedGroup of FlxObjects to store player and enemies which could land on platform. Everything is synced up, but collision logic/variables have NOT been implemented.

CheezitBandit committed:
Sprite touching platform now gets added to group of separate class, which tracks offset and contact with platform. DOES NOT get removed/reset offset yet, but stays in contact.

CheezitBandit committed:
Final (probably) submission for moving platforms - separated into separate class, flxgroup tracks all sprites and checks if they're in contact
Only issue remaining is a slight jump as the platform changes from up to down.

SamBumgardner commented:
I'll test the multi-sprite functionality today, and let you know if the branch is ready to merge. After that, you can resolve merge conflicts and we'll be good to go!

SamBumgardner commented:
Okay, I found that your code crashes when multiple objects leave the platform. This happens because you remove objects from your list while iterating through it, so you end up trying access an invalid index by the end of the loop.

I also found out that the player hopping off of the top of the platform is because the player's acceleration wasn't being set to the `heavy` variable when it lands on the platform.

In the process of figuring out what was going wrong, I fixed the issues. I'll do another branch and PR to your branch thing to save everyone some trouble.

SamBumgardner committed:
Changed the touchingSprites removal strategy to prevent crash.

When removing objects from a list you're iterating over, you need to make sure to move through
the list from back to front. The code here executes that logic.

SamBumgardner committed:
Fixed sprites hopping off of a platform when it transitions to moving down.

This is done by re-implementing the old "heavy gravity while touching platform" logic.
When the player enters contact with the platform, its y acceleration increases dramatically.
(done inside platform code)

When the player enters the air, its air acceleration is reset to normal values.
(done inside player state code).

SamBumgardner committed:
Removed extra update call in Platforms.

The objects in the touchingSprites group aren't actually added to the
game's current state, and their update function doesn't really do anything
important, so this isn't really necessary. I'll also comment on this in the
PR review.

SamBumgardner committed:
Changed the map to collide against all sprites, rather than just the player.

SamBumgardner commented:
Okay, PR has been made. After mine gets merged in, I'm fine with this getting merged too. 

There are some things I'd like to see changed (I'll put comments on them shortly), but I think you've done enough work on this thing already. And it functions properly, which was the original goal!

SamBumgardner commented on source/PlatformTracker.hx:
```diff
@@ -0,0 +1,36 @@
+ * @author Jess Geiger
+ */
+
+ class PlatformTracker extends Player
```
OPTIONAL FIX: 

I don't think this PlatformTracker class really needs to extend Player.
It doesn't use any of player's variables or methods in any situation. It doesn't need a hitbox or a graphic either, which are things that come with inheriting from Player (which inherited from FlxSprite). 
This really is just a little bundle of data, so I don't think you need it to inherit from anything.

CheezitBandit replied:
You're right, it was unnecessary. I was copying the class declarations of the other classes and assumed that in Haxe, everything had to inherit from one of the top classes (FlxSprite, FlxState, etc). Rookie mistake, and it has been corrected.

SamBumgardner commented on source/PlatformTracker.hx:
```diff
@@ -0,0 +1,36 @@
+    public var trackedOffsetX:Float;
+    public var baseObj:FlxObject;
+
+    public function new(?obj:FlxObject, ?offsetX:Float = 0)
```
Could use basic documentation.
See the Player class for examples.

CheezitBandit replied:
Done

CheezitBandit replied:
/** 
     *  Initialize the new PlatformTracker object and set its variables to the requested initial values.
     *  
     *  @param trackedOffsetX is the difference in the x position of the platform
     *  and the x position of the object when it first touches the platform
     *  @param baseObj is the base player/enemy sprite object which is built upon
     *  in this class
     */

SamBumgardner commented on source/PlatformTracker.hx:
```diff
@@ -0,0 +1,36 @@
+        baseObj = obj;    
+    }
+
+    public function returnOffset():Float
```
Basic documentation needed.

SamBumgardner replied:
Or you could get rid of it entirely, since baseObj is a public variable and can be accessed without this function.

CheezitBandit replied:
You're right, that was unnecessary code. It has been removed.

SamBumgardner commented on source/PlatformTracker.hx:
```diff
@@ -0,0 +1,36 @@
+        return trackedOffsetX;
+    }
+
+    public function returnBase():FlxObject
```
Basic documentation needed.

Or remove it and just have other stuff access baseObj directly, up to you.

CheezitBandit replied:
It's been removed in favor of directly accessing the object.

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,178 @@
+    public var touchingSprites:FlxTypedGroup<PlatformTracker> = new FlxTypedGroup<PlatformTracker>();
+    private var countTouching:Int = 0;
+
+    public function new(?X:Float=0, ?Y:Float=0, ?W:Int=0, ?L:Float=0, ?R:Float=0, ?U:Float=0, ?D:Float=0, ?trackPlayer:Player, ?SimpleGraphic:FlxGraphicAsset) 
```
Missing basic documentation.

CheezitBandit replied:
Done

CheezitBandit replied:
/**
     *  Create object and initialize variables based on PlayState declarations
     *  
     *  @param   trackPlayer is used to make the player accessible to the platform class
     */

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    // Add or update sprites colliding with platform
+    public function platformObjects(elapsed:Float, player:FlxObject, platform:Platforms):Void
```
Missing basic documentation.

CheezitBandit replied:
Done

CheezitBandit replied:
/**
     *  Add or update sprites colliding with the platform. Add the sprite to the touching platform group if
     *  the group is empty. If it is not empty, iterate through until the sprite is either found (in which
     *  case update its positioning) or add it to the list if it is not already present.
     *  
     *  @param   elapsed is the time passed since last update
     *  @param   player is the sprite object
     *  @param   platform is the platform which the sprite object is touching
     */

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,178 @@
+    public var offsetX:Float = 0;
+    public var inContact:Bool = false;
+    private var platformVelocity:Int = 40;
+    public var touchingSprites:FlxTypedGroup<PlatformTracker> = new FlxTypedGroup<PlatformTracker>();
```
OPTIONAL FIX: 

FlxTypedGroups are mostly useful because of their recycle and update functions, and the fact that adding it to a state is like adding all of its members to the state in one go.

In this class, the touchingSprites group pretty much just used as an array, so I think you might as well just change its type to `Array<PlatformTracker>` and adjust all of the spots its used in to use array methods directly instead of using the `touchingSprites.members` array.

CheezitBandit replied:
You're right, that was a much simpler way to implement that. I thought since the sprites group added to PlayState.hx was a Typed Group that it would be easiest to keep the touchingSprites group as the same, but an array works much better.

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,178 @@
+	}
+
+    //Movement as requested by user parameters
+    public function movement():Void
```
Missing basic documentation.

CheezitBandit replied:
Done

CheezitBandit replied:
/**
     *  Movement as requested by user parameters. When the object hits the maximum position in 
     *  one direction, it is directed in the other direction. Corresponding booleans are changed
     *  as well.
     *  
     */

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    // Remove sprites NOT colliding with platform
+    private function notPlatformObjects(elapsed:Float, sprite:FlxObject, platform:Platforms):Void
```
Missing basic documentation.

CheezitBandit replied:
Done

CheezitBandit replied:
/**
     *  Remove sprites NOT colliding with the platform. If objects are touching the platform, check that 
     *  this sprite is not one of them. If the object is found in the list of sprites touching, 
     *  remove it from the group and set its y acceleration back to Gravity as declared in PlayState.hx
     *  
     *  @param   elapsed is the time passed since last update
     *  @param   sprite is the sprite object not touching the platform
     *  @param   platform is the platform not being touched by the sprite
     *  @param   originalCount is the number of sprites in the group before any changes are made; used to 
     *  prevent changes to loop duration
     */

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,178 @@
+	
+
+
+    public function platformUpdate(elapsed:Float, objects:FlxTypedGroup<FlxObject>, platform:Platforms):Void
```
Missing basic documentation.

CheezitBandit replied:
Done

CheezitBandit replied:
/**
     *  Function called on every update in PlayState.hx. Iterates through every sprite object on the map and 
     *  checks each to see if it is either touching or not touching the platform
     *  
     *  @param   objects is the list of all sprites present
     *  @param   platform is the platform with which sprites are checked for contact
     */

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,178 @@
+    }
+
+
+    public override function update(elapsed:Float):Void
```
Technically missing basic documentation, but I'm okay with letting this slide because it doesn't need much explaining.

CheezitBandit replied:
Got it. Sorry, I should have looked at the documentation in Player.hx, I completely missed the standard I needed to meet. It should (hopefully) all be corrected now!

CheezitBandit replied:
/**
     *  Call movement function to determine movement of platform and update the platform's position
     *  
     */

SamBumgardner commented on source/Platforms.hx:
```diff
@@ -0,0 +1,178 @@
+    {
+        for (i in 0...objects.length)
+        {
+            if (FlxG.collide(platform, objects.members[i])) {
```
OPTIONAL FIX:

Colliding objects against one another one-by-one is less efficient than colliding two groups against one another all at once.

What I was thinking was that you have a single overlap resolution function that does collision separation and sets some `has_collided` boolean value in its PlatformTracker to `true`. Then you can collide all objects against the platform in a single call to FlxG.overlap. 

Doing stuff like moving tracked object's X coordinate or removing PlatformTracker objects when things didn't collide with the platform this frame can all be done in a function in the platform's update.

The platform will loop through its tracked objects and look at each one's `has_collided` then decide if it should do the logic that's currently in your platformObjects or your notPlatformObjects functions.

But like I said, this change is optional. We probably won't notice the performance distance with how small the game is overall.

CheezitBandit replied:
I don't think I'll implement this... I'll look at the other changes you suggested, but this would be last on my list of things to change. You're right though, in a larger game it would definitely slow down. It wouldn't scale well.

SamBumgardner reviewed (changes requested):
Okay, here's my last review. Definitely put in some documentation in the requested spots (make sure to do the `/**` style found in `Player.hx`) and fix any of the optional things that you'd like. 

After you feel like it's done, let me know and we'll likely be good to merge.

CheezitBandit committed:
Merge pull request #2 from SamBumgardner/fix-multiple-collisions

Fix multiple collisions

CheezitBandit commented on source/PlatformTracker.hx:
```diff
@@ -0,0 +1,36 @@
+    public var trackedOffsetX:Float;
+    public var baseObj:FlxObject;
+
+    public function new(?obj:FlxObject, ?offsetX:Float = 0)
```
/**
     *  Create object and initialize variables based on PlayState declarations
     *  
     *  @param   trackPlayer is used to make the player accessible to the platform class
     */

CheezitBandit committed:
Mostly editorial changes to comments as requested in SamBumgardner/csc-303-platformer pull request #6

CheezitBandit committed:
Converted touchingSprite FlxTypedGroup to Array; removed unnecessary inheritance on class PlatformTracker

CheezitBandit committed:
Merge branch 'master' into Collision_Branch

CheezitBandit commented:
I've made the changes you suggested and resolved the merge conflicts with PlayState.hx between Collision_Branch and Master. Please let me know if there are any further changes that need to be made.

SamBumgardner commented:
Excellent work on this branch, I'll merge it in now!
